File: Data Structures & Algorithms/sliding-window-maximum/submission-0.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {

        std::map<int, int> freqs;
        vector<int> ans;

        for (int i = 0; i < k; i++) {
            int num = nums[i];
            if (freqs.find(num) != freqs.end()) {
                freqs[num]++;
            } else {
                freqs[num] = 1;
            }
        }

        for(int i = k; i < nums.size(); i++) {
            ans.push_back((--freqs.end())->first);

            freqs[nums[i-k]] -= 1;
            int num = nums[i];
            if (freqs.find(num) != freqs.end()) {
                freqs[num]++;
            } else {
                freqs[num] = 1;
            }

            if (freqs[nums[i-k]] == 0){
                freqs.erase(nums[i-k]);
            }
        }
        ans.push_back((--freqs.end())->first);

        return ans;
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = 1
        zero_count = 0
        for num in nums:
            if (num != 0):
                res *= num
            else:
                zero_count += 1

        ans = []
        for num in nums:
            if zero_count == 0:
                if num != 0:
                    ans.append(int(res/num))
            elif zero_count >= 2:
                ans.append(0)
            else :
                if num == 0:
                    ans.append(res)
                else:
                    ans.append(0)
        return ans